teaching-inheritance/inheritance.md

```diff
@@ -0,0 +1,165 @@
+  - The book does not provide any context for how this code will be used. I would argue this is is a major oversight. Writing code in a vacuum is fine for experienced engineers, but in my experience novices benefit from concrete examples of how code will be used from the "outside". With inheritance in particular, this would demonstrate how polymorphism is useful.
+- There is no mention of the idea of extending a class implemented by a framework.
+
+In general _Building Java Programs_ is excellent, and I have a tremendous amount of respect for Reges and Stepp. It certainly did a good job of preparing my students for the AP CS exam. However, it does not introduce inheritance as it is used in the real world, particularly by novice engineers. As far as I can tell this is typical of introductory CS courses - certainly my undergraduate education at Purdue followed a similar pattern.
+
+### Design Textbooks
+
+There is another type of text that addresses inheritance: books on software design. Famous resources like _Practical Object-Oriented Design: An Agile Primer Using Ruby_ (POODR) by Sandi Metz, or _Design Patterns_ (the "Gang of Four" book) by Gamma, Helm, Johnson and Vlissides address software design more generally, employing inheritance as one tool among many.
+
+However, these books are targeted at experienced engineers trying to up their game, not at novices learning their chosen language for the first time. Moreover they discuss design from a ground-up perspective, whereas an engineer beginning their career is likely to build on the shoulders of giants, extending existing designs rather than inventing new ones.
+
+An ideal curriculum would bridge the gap between these two approaches, introducing both syntax and common inheritance idioms, and getting new engineers used to the work of extending an existing class.
+
+## Ada's Approach to Inheritance
+
+The instructional team at Ada has been unhappy with our approach to inheritance for a while now, but haven't quite known what to do about it. Now that we've done some research and formalized our engineering and pedagogical intuition, here's the approach we've come up with:
+
+- Simple examples and accessible metaphors are fine for [introducing syntax and semantics](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/object-inheritance.md), though as Reges and Stepp demonstrate they don't have to be completely unrealistic.
+- Common idioms like abstract classes and the template method pattern should be introduced as soon as the basic syntax is understood. **TODO CURRICULUM LINK**
+- Students' first serious inheritance project should involve extending an existing superclass.
+  - This matches the way inheritance is used in the real world, and makes the benefits (not having to re-write a bunch of code) immediately clear.
+  - Instructors would provide the following scaffolding:
+    - Superclass implementation
+    - Driver code demonstrating polymorphism
```

Comment:
This is the first time that polymorphism is mentioned. Can we add some more on polymorphism in thesection earlier related to why inheritanceis hard to teach? E.g. nuances like language support, or lack thereof for compile time polymorphism. How do we teach inheritance and polymorphism with the complete benefits and boundaries of one language and yet prepare our student engineers to take on any programming language in the future?